enable pmt on block

elliothllm opened the pull request:
### Description
Enable PMT via block in the chainspec. This allows for swapping over from SMT to PMT. Also enable simultaneous building of PMT and SMT.

The idea here is to sync an RPC node using the flag `zkevm.simultaneous-pmt-and-smt: true` to build the PMT in the DB. Then promote the RPC node to a Sequencer node updating the chainspec with the block you paused at. Then the new Sequencer will use PMT.

It is not possible to unwind past the point of swap over currently.

**Chainspec**
`"pmtEnabledBlock": 0`

**Flags**
`zkevm.simultaneous-pmt-and-smt: true`

elliothllm committed:
enable pmt on block

pmt block enablement fix tlags after rebase

pmt block enablement fix failing tests

hexoscott commented on turbo/jsonrpc/zkevm_api.go:
```diff
@@ -1094,6 +1099,15 @@ func (api *ZkEvmAPIImpl) getBlockRangeWitness(ctx context.Context, db kv.RoDB, s
 		return nil, fmt.Errorf("start block number must be less than or equal to end block number, start=%d end=%d", blockNr, endBlockNr)
 	}
 
+	cc, err := api.ethApi.chainConfig(ctx, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.IsPmtEnabled(blockNr) {
+		return nil, ErrStateTrieUnsupported
+	}
+
```
We could shunt this up higher in the function so that we can exit earlier in this case.

hexoscott replied:
Scratch that, very little gain

hexoscott commented on turbo/jsonrpc/zkevm_api.go:
```diff
@@ -1194,6 +1202,15 @@ func (api *ZkEvmAPIImpl) GetProverInput(ctx context.Context, batchNumber uint64,
 		return nil, err
 	}
 
+	cc, err := api.ethApi.chainConfig(ctx, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	if cc.IsPmtEnabled(blockNumbers[0]) {
+		return nil, errors.New("state trie does not support witness")
+	}
+
```
We can probably yoink this block higher up as well perhaps?  One thing we might want to consider as well is using a `sync.Once` so that we only perform this lookup once during runtime.  The chain config won't change after app startup so we can do this read and check once and store it on the `api` pointer from the method call.

hexoscott replied:
Same here, lets leave this as is

hexoscott reviewed (changes requested):
One really small change request from the RPC side of things

hexoscott committed:
use execution progress to determine PMT rather than stage height

if we use the inters stage height then we will use the SMT instead of
PMT when we wake the network up after defining the override block heights
and times

hexoscott committed:
fix for shanghai -> prague sender recovery

sending 0 means that prague signer can't be used if prague was enabled
beyond genesis.

IvanBelyakoff commented on core/types/genesis.go:
```diff
@@ -70,7 +70,17 @@ type Genesis struct {
 
 	// Zkevm fields
 	HonourChainspec bool `json:"honourChainspec,omitempty"` // Used in normalcy mode
```
I can't remember any case when we want normalcy but we don't want HonourChainspec. When I was load testing type2 and type1, that was just an extra burden to edit in the config. Maybe we just remove it and if normalcy, we honour chainspec.

hexoscott replied:
Lets keep it there for now and we can remove it in a subsequent change and deprecate the flag.

IvanBelyakoff commented on erigon-lib/chain/chain_config.go:
```diff
@@ -703,6 +716,7 @@ type Rules struct {
 	IsAura                                                                                                                                               bool
 	IsNormalcy                                                                                                                                           bool
+	IsPmtEnabled                                                                                                                                         bool
 	IsType1                                                                                                                                              bool
```
Should we remove IsType1 field now? This one is now totally composable of PmtEnabled and Normalcy and we have a public function for it. I believe this field is never used

hexoscott replied:
I'd argue for consistency we can keep this, we do the same thing for other fields by just calling the underlying public function i.e. `IsForkID13Durian:     c.IsForkID13Durian(num)`

IvanBelyakoff reviewed (dismissed):
Awesome job!
I left a couple of minor comments, maybe we could address them in a follow up PR.
One thing I am curious about is if we need any SMT data pruning after we switched to PMT?

hexoscott committed:
sequencer to use v2 SMT algo

hexoscott committed:
handle upgrades from PP to type 1 networks

As the fork history isn't there when we migrate we write it but
need to support handling the fork upgrade in the batch the sequencer
is about to produce

hexoscott committed:
updating london tests to remove legacy flags and price hints

hexoscott committed:
test update around PP -> PP migrations between type 2 and type 1

hexoscott committed:
adding a comment around pp type 2 to type 1 migration for fork checking
